app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionRepositoryImpl.kt

```diff
@@ -0,0 +1,144 @@
+package org.dhis2.usescases.programStageSelection
+
+import androidx.annotation.VisibleForTesting
+import io.reactivex.Flowable
+import io.reactivex.Single
+import io.reactivex.functions.Function5
+import java.util.HashMap
+import org.dhis2.Bindings.toRuleAttributeValue
+import org.dhis2.data.forms.RulesRepository
+import org.dhis2.utils.EventCreationType
+import org.dhis2.utils.Result
+import org.hisp.dhis.android.core.D2
+import org.hisp.dhis.android.core.enrollment.Enrollment
+import org.hisp.dhis.android.core.event.Event
+import org.hisp.dhis.android.core.program.ProgramStage
+import org.hisp.dhis.android.core.program.ProgramTrackedEntityAttribute
+import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue
+import org.hisp.dhis.rules.RuleEngine
+import org.hisp.dhis.rules.RuleEngineContext
+import org.hisp.dhis.rules.RuleExpressionEvaluator
+import org.hisp.dhis.rules.models.Rule
+import org.hisp.dhis.rules.models.RuleEffect
+import org.hisp.dhis.rules.models.RuleEnrollment
+import org.hisp.dhis.rules.models.RuleEvent
+import org.hisp.dhis.rules.models.RuleVariable
+import org.hisp.dhis.rules.models.TriggerEnvironment
+
+class ProgramStageSelectionRepositoryImpl internal constructor(
+    evaluator: RuleExpressionEvaluator,
+    rulesRepository: RulesRepository,
+    private val programUid: String,
+    private val enrollmentUid: String,
+    private val eventCreationType: String,
+    private val d2: D2
+) : ProgramStageSelectionRepository {
+
+    private val cachedRuleEngineFlowable: Flowable<RuleEngine>
+
+    @VisibleForTesting
+    fun ruleEnrollment(): Flowable<RuleEnrollment> =
```

Comment:
All methods in Kotlin are public by default. There is no need for the annotation in here. 👍